submissions/angels_unit_tests/angels_unit_tests.md

```diff
@@ -0,0 +1,55 @@
+
+With these 4 scripts, everything is in place to launch factorio with a certain configuration of the latest mods and their mod settings. It was only a matter of telling factorio to launch over and over with different mod settings. Assuming an ingame mod (Angel's Dev - Unit Tests in this case) telling the testing infrastructure that the testing is over, it can shut down Factorio and launch the next configuration to test.
+
+## The last piece of the puzzle
+
+A week later, on the 8th of June 2021 I finished writing [the most basic unit test](https://github.com/Arch666Angel/mods/pull/634/commits/1435fbd2f1134b08307406e21fdd657f1cafa019) which did nothing more than reporting which mods are active ingame. After the unit test was finished logging the results, the [main script](https://github.com/Arch666Angel/mods/pull/634/commits/64ea68023f19c2505aa3cfbb7642f13cf7f70fbd) could launch the next configuration.
+
+As any developer will know, this minimal viable product (MVP) could still use some polishing. After some extensive use, on the 22nd of August 2021, I added two more features, of which the first added a separate file containing [all the test configurations](https://github.com/Arch666Angel/mods/commit/27b899d43b1db04d99cc8c8086cf7ece83376419#diff-f7f9cb110df5ea99bfdecacc536b281dba523acc5418d1e9774fc803f7846348), such that it becomes easier to add new configurations as needed. The second feature added the ability to [log the test results to a file](https://github.com/Arch666Angel/mods/commit/a5732a52c5dfdcd6d9c5f4d89ad699ea9bd90bda#diff-f7f9cb110df5ea99bfdecacc536b281dba523acc5418d1e9774fc803f7846348). Doing this the new test results could be compared with the old test results by using some external tools such as [ExamDiff Pro](https://www.prestosoft.com/edp_examdiffpro.asp). This is mainly to verify that some rapported issues are resolved (lines missing in the new test results), without introducing new issues (lines added in the new test results).
+
+## The result
+
+Since the use of the minimal viable product, we released already an update. After two weeks of waiting for major game breaking bugs, we were surprised at finding that we had close to no breaking bugs. I personally have a new 'rule' while maintaining Angel's mods; if a bug is found, first create an appropriate unit test to catch this bug, before resolving the issue. My personal feedback, it takes quite some additional time to run all the unit tests, but that is easily planned by running the unit tests while taking a coffee break.
```

Comment:
Are you doing any runtime tests? Like simulating players running, or some scripts or something? Might be worth a mention.

Reply:
Fixed in https://github.com/AlternativeFFFF/Alt-F4/pull/551/commits/0d2bc1d8be83f76d557f22d7d39ba53b96e0aa9f